src/registry2.0/portimage.tcl

```diff
@@ -348,9 +348,18 @@ proc extract_archive_to_tmpdir {location} {
             t(ar|bz|lz|xz|gz) {
                 set tar "tar"
                 if {[catch {set tar [macports::findBinary $tar ${macports::autoconf::tar_path}]} errmsg] == 0} {
-                    ui_debug "Using $tar"
-                    set unarchive.cmd "$tar"
-                    set unarchive.pre_args {-xvpf}
+                    # Opportunistic HFS compression. bsdtar will automatically
+                    # ignore if on non-HFS filesystem.
+                    if {![catch {macports::binaryInPath bsdtar}] &&
+                        ![catch {exec bsdtar -x --hfsCompression < /dev/null >& /dev/null}]} {
```

Comment:
Might as well just check `[file executable ${prefix}/bin/bsdtar]` (and not bother checking if --hfsCompression is supported) since that's the only place a suitably new bsdtar will be. If the next macOS release ships one, we can check for it once at configure time.

Reply:
That's exactly what I'm doing in my version of this patch as well as in a couple of ports that unpack huge source trees. Works perfectly, as long as the target filesystem supports compression.  I don't know how bsdtar handles pre-compression HFS versions but it will raise an error when unpacking to a non-HFS filesystem. That doesn't just concern older OS X versions; there is no obligation that I know of to install MacPorts to an Apple filesystem (all of ${prefix} could be on a ZFS dataset, for instance).
/Methinks that for the current application it should suffice to provide an option in macports.conf, possible setting it appropriately during the initial install. This is a local configuration aspect that won't change often. Or, as I do in my local `compress_workdir` PG: query an env. variable to opt in or out of the feature.

> What happens if ${portdbpath}/software is on a different volume than the rest of ${prefix}?

That's going to depend on how you move the unpacked (MPExtract) directory to its destination. Currently compression is lost (my /Applications/Macports is on a different volume than ${prefix}) but that can be avoided if the move is handled through `ditto` or `${prefix}/rsync`.

Reply:
>What happens on Leopard and Tiger?

I don't know with absolute certainty (haven't tested), but according to https://github.com/aosm/xnu/commit/d7b24a84012524e9dfcec55e4a6f2a5ff1f00199 the UF_COMPRESSED flag was not defined prior to 10.10, so the code which turns on the compress steps (but *not* the code that permits the --hfsCompression flag to be parsed) is disabled, so it be comes a no-op: https://github.com/libarchive/libarchive/blob/c10875db78b04fd730bb08245b7032baba906815/libarchive/archive_write_disk_posix.c#L634

> Might as well just check [file executable ${prefix}/bin/bsdtar] (and not bother checking if --hfsCompression is supported)

My thought here was that a user might install bsdtar (somewhere else in their PATH) on their own if they want to make sure it is always available; a corner case to be sure... I'm fine either way. It would be quicker to not have to check the flag, but it's also not slow. (Only done once per extraction.)

>I don't know how bsdtar handles pre-compression HFS versions but it will raise an error when unpacking to a non-HFS filesystem. 

No, it (bsdtar) silently disables the compression step. See the email discussion chain and the libarchive source.

>>What happens if ${portdbpath}/software is on a different volume than the rest of ${prefix}

> That's going to depend on how you move the unpacked (MPExtract) directory to its destination. Currently compression is lost (my /Applications/Macports is on a different volume than ${prefix}) but that can be avoided if the move is handled through ditto or ${prefix}/rsync.

Good point. That is a separate step and can be addressed separately. This change won't break that step in that case, but it will be unnecessarily slower for the (wasted) compression action. A check could certainly be added to see if our extract location and expected install locations ($prefix and /Applications/MacPorts) are on different mount points; but I suspect many (most?) users aren't in that situation.